### ISSUE TITLE
Workers' connection to scheduler breaks when using Dask (with dashboard) and JupyterLab on interactive compute node of a SGE HPC

### ISSUE THREAD

**Issue by lukeconibear** [dask/distributed#3857]
**What happened**:
I'm trying to use Dask and JupyterLab on an _interactive compute node_ of a SGE HPC. The general approach I'm using follows the Pangeo deployment of JupyterLab and Dask jobqueue, detailed [here](https://pangeo.io/setup_guides/hpc.html#deploy-option-1-jupyter-dask-jobqueue). When I execute `.compute()`, JupyterLab gets a server connection error, the workers are killed, and the interactive job on the HPC is killed. This issue does not occur if I independently turn off the Dask dashboard using `extra=['--no-dashboard']`, if I run with a small number of workers (e.g. 5), or if I run JupyterLab with the Dask dashboard on a _login node_ of the same HPC. The output diagnostic from the workers return that the connection to the scheduler was broken or that the nanny fails to start. I previously posted this issue on a closed GitHub issue for Dask jobqueue [here](https://github.com/dask/dask-jobqueue/issues/391#issuecomment-634683278), and followed a suggestion to see if someone here has any ideas. I have contacted my local cluster IT to see whether they have network restrictions on the compute nodes that could impact websockets, and I'm awaiting a response.

**What you expected to happen**:
The workers to remain connected to the scheduler and the nanny starts, whilst using JupyterLab with the Dask dashboard on an interactive compute node of a SGE HPC.

**Minimal Complete Verifiable Example**:
```python
from dask_jobqueue import SGECluster
from dask.distributed import Client
import dask.array as da

cluster = SGECluster()
client = Client(cluster)

cluster.scale(jobs=50)

x = da.random.random((10000, 10000), chunks=(1000, 1000))
y = x + x.T

y.compute() # server connection error occurs here
```

**Anything else we need to know?**:
Below is example output diagnostic from the workers that either lose their connection to the scheduler or where the nanny fails to start.

```bash
distributed.nanny - INFO -         Start Nanny at: 'tcp://10.12.169.175:39497'
distributed.worker - INFO -       Start worker at:  tcp://10.12.169.175:46064
distributed.worker - INFO -          Listening to:  tcp://10.12.169.175:46064
distributed.worker - INFO -          dashboard at:        10.12.169.175:36365
distributed.worker - INFO - Waiting to connect to:   tcp://10.12.169.99:45552
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -               Threads:                          1
distributed.worker - INFO -                Memory:                 1000.00 MB
distributed.worker - INFO -       Local Directory: /nobackup/earlacoa/dask/scripts/dask-worker-space/worker-7d91wp5m
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -         Registered to:   tcp://10.12.169.99:45552
distributed.worker - INFO - -------------------------------------------------
distributed.core - INFO - Starting established connection
distributed.worker - INFO - Connection to scheduler broken.  Reconnecting...
tornado.application - ERROR - Exception in callback <function Worker.__init__.<locals>.<lambda> at 0x2b4993bd60d0>
Traceback (most recent call last):
  File "/nobackup/earlacoa/miniconda3/envs/pangeo/lib/python3.6/site-packages/tornado/ioloop.py", line 907, in _run
    return self.callback()
  File "/nobackup/earlacoa/miniconda3/envs/pangeo/lib/python3.6/site-packages/distributed/worker.py", line 644, in <lambda>
    lambda: self.batched_stream.send({"op": "keep-alive"}), 60000,
  File "/nobackup/earlacoa/miniconda3/envs/pangeo/lib/python3.6/site-packages/distributed/batched.py", line 117, in send
    raise CommClosedError
distributed.comm.core.CommClosedError
distributed.worker - INFO - Stopping worker at tcp://10.12.169.175:46064
distributed.nanny - INFO - Worker closed
distributed.nanny - INFO - Closing Nanny at 'tcp://10.12.169.175:39497'
distributed.dask_worker - INFO - End worker
```

```bash
distributed.nanny - INFO -         Start Nanny at: 'tcp://10.12.169.41:36483'
distributed.worker - INFO -       Start worker at:   tcp://10.12.169.41:33432
distributed.worker - INFO -          Listening to:   tcp://10.12.169.41:33432
distributed.worker - INFO -          dashboard at:         10.12.169.41:45407
distributed.worker - INFO - Waiting to connect to:   tcp://10.12.169.99:45552
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -               Threads:                          1
distributed.worker - INFO -                Memory:                 1000.00 MB
distributed.worker - INFO -       Local Directory: /nobackup/earlacoa/dask/scripts/dask-worker-space/worker-f5q2x92l
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO - Waiting to connect to:   tcp://10.12.169.99:45552
distributed.worker - INFO - Waiting to connect to:   tcp://10.12.169.99:45552
distributed.worker - INFO - Waiting to connect to:   tcp://10.12.169.99:45552
distributed.worker - INFO - Waiting to connect to:   tcp://10.12.169.99:45552
distributed.worker - INFO - Waiting to connect to:   tcp://10.12.169.99:45552
distributed.nanny - INFO - Closing Nanny at 'tcp://10.12.169.41:36483'
distributed.worker - INFO - Stopping worker at tcp://10.12.169.41:33432
distributed.worker - INFO - Closed worker has not yet started: None
distributed.nanny - WARNING - Worker process still alive after 0 seconds, killing
distributed.dask_worker - INFO - End worker
Traceback (most recent call last):
  File "/nobackup/earlacoa/miniconda3/envs/pangeo/lib/python3.6/site-packages/distributed/core.py", line 226, in _
    await asyncio.wait_for(self.start(), timeout=timeout)
  File "/nobackup/earlacoa/miniconda3/envs/pangeo/lib/python3.6/asyncio/tasks.py", line 362, in wait_for
    raise futures.TimeoutError()
concurrent.futures._base.TimeoutError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/nobackup/earlacoa/miniconda3/envs/pangeo/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/nobackup/earlacoa/miniconda3/envs/pangeo/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/nobackup/earlacoa/miniconda3/envs/pangeo/lib/python3.6/site-packages/distributed/cli/dask_worker.py", line 452, in <module>
    go()
  File "/nobackup/earlacoa/miniconda3/envs/pangeo/lib/python3.6/site-packages/distributed/cli/dask_worker.py", line 448, in go
    main()
  File "/nobackup/earlacoa/miniconda3/envs/pangeo/lib/python3.6/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/nobackup/earlacoa/miniconda3/envs/pangeo/lib/python3.6/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/nobackup/earlacoa/miniconda3/envs/pangeo/lib/python3.6/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/nobackup/earlacoa/miniconda3/envs/pangeo/lib/python3.6/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/nobackup/earlacoa/miniconda3/envs/pangeo/lib/python3.6/site-packages/distributed/cli/dask_worker.py", line 434, in main
    loop.run_sync(run)
  File "/nobackup/earlacoa/miniconda3/envs/pangeo/lib/python3.6/site-packages/tornado/ioloop.py", line 532, in run_sync
    return future_cell[0].result()
  File "/nobackup/earlacoa/miniconda3/envs/pangeo/lib/python3.6/site-packages/distributed/cli/dask_worker.py", line 428, in run
    await asyncio.gather(*nannies)
  File "/nobackup/earlacoa/miniconda3/envs/pangeo/lib/python3.6/asyncio/tasks.py", line 537, in _wrap_awaitable
    return (yield from awaitable.__await__())
  File "/nobackup/earlacoa/miniconda3/envs/pangeo/lib/python3.6/site-packages/distributed/core.py", line 232, in _
    type(self).__name__, timeout
concurrent.futures._base.TimeoutError: Nanny failed to start in 60 seconds
```

**Environment**:
```bash
dask                      2.17.0                     py_0    conda-forge
dask-core                 2.17.0                     py_0    conda-forge
dask-jobqueue             0.7.1                      py_0    conda-forge
dask-labextension         2.0.2                    pypi_0    pypi
jupyter                   1.0.0                      py_2    conda-forge
jupyter-server-proxy      1.5.0                    pypi_0    pypi
jupyter_client            6.1.3                      py_0    conda-forge
jupyter_console           6.1.0                      py_1    conda-forge
jupyter_core              4.6.3            py36h9f0ad1d_1    conda-forge
jupyterlab                2.1.3                      py_0    conda-forge
jupyterlab_server         1.1.5                      py_0    conda-forge
tornado                   6.0.4            py36h8c4c3a4_1    conda-forge
python                    3.6.10          h8356626_1011_cpython    conda-forge
```

**Comment by lesteve** [comment]
I helped @lukeconibear and recommended that he post here for more feed-back because some people may have seen a similar issue before. @lukeconibear can you also post the error you get in your Jupyter session with the websocket error?

To me it looks like for some reason the Dask scheduler (that lives in the Jupyter kernel) dies after some time, so I guess if someone has some suggestions as to how to investigate, that would be more than welcome!

Something else worth trying is to run the same code in a IPython console to see whether the problem is more related to JupyterLab or the JupyterLab Dask extension.

**Comment by lukeconibear** [comment]
Sure, the websocket error is below:

```bash
(pangeo) [earlacoa@d8s0b3.arc4 dask]$ jupyter lab --no-browser --ip=`hostname` --port=8888
[I 14:23:01.735 LabApp] JupyterLab extension loaded from /nobackup/earlacoa/miniconda3/envs/pangeo/lib/python3.6/site-packages/jupyterlab
[I 14:23:01.735 LabApp] JupyterLab application directory is /nobackup/earlacoa/miniconda3/envs/pangeo/share/jupyter/lab
[I 14:23:01.770 LabApp] Serving notebooks from local directory: /nobackup/earlacoa/dask
[I 14:23:01.771 LabApp] The Jupyter Notebook is running at:
[I 14:23:01.771 LabApp] http://d8s0b3.arc4.leeds.ac.uk:8888/
[I 14:23:01.771 LabApp] Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).
[I 14:23:56.835 LabApp] Build is up to date
TermSocket.open: 1
TermSocket.open: Opened 1
/Websocket closed
\[I 14:24:05.747 LabApp] Kernel started: 6e8bb670-4631-4a5e-bdb3-bd79c29f1d69
/TermSocket.open: 2
TermSocket.open: Opened 2
[I 14:30:05.138 LabApp] Client sent subprotocols: ['bokeh', 'eyJzZXNzaW9uX2lkIjogIjdGVnBMdDdnaGZaQ2JVckowN25CMkc1a3BuWWdqaENzQVBOZXR4OWVVRWVmIiwgInNlc3Npb25fZXhwaXJ5IjogMTU5MDU4OTg4MiwgImhlYWRlcnMiOiB7Ikhvc3QiOiAibG9jYWxob3N0Ojg4ODgiLCAiQ29ubmVjdGlvbiI6ICJrZWVwLWFsaXZlIiwgIkNhY2hlLUNvbnRyb2wiOiAibWF4LWFnZT0wIiwgIlVwZ3JhZGUtSW5zZWN1cmUtUmVxdWVzdHMiOiAiMSIsICJVc2VyLUFnZW50IjogIk1vemlsbGEvNS4wIChXaW5kb3dzIE5UIDEwLjA7IFdpbjY0OyB4NjQpIEFwcGxlV2ViS2l0LzUzNy4zNiAoS0hUTUwsIGxpa2UgR2Vja28pIENocm9tZS84MS4wLjQwNDQuMTM4IFNhZmFyaS81MzcuMzYiLCAiQWNjZXB0IjogInRleHQvaHRtbCxhcHBsaWNhdGlvbi94aHRtbCt4bWwsYXBwbGljYXRpb24veG1sO3E9MC45LGltYWdlL3dlYnAsaW1hZ2UvYXBuZywqLyo7cT0wLjgsYXBwbGljYXRpb24vc2lnbmVkLWV4Y2hhbmdlO3Y9YjM7cT0wLjkiLCAiU2VjLUZldGNoLVNpdGUiOiAiY3Jvc3Mtc2l0ZSIsICJTZWMtRmV0Y2gtTW9kZSI6ICJuYXZpZ2F0ZSIsICJTZWMtRmV0Y2gtVXNlciI6ICI_MSIsICJTZWMtRmV0Y2gtRGVzdCI6ICJkb2N1bWVudCIsICJBY2NlcHQtRW5jb2RpbmciOiAiZ3ppcCIsICJBY2NlcHQtTGFuZ3VhZ2UiOiAiZW4tR0IsZW4tVVM7cT0wLjksZW47cT0wLjgiLCAiQ29va2llIjogIl94c3JmPTJ8OGI5MjJlMmN8YTA3OTZkZWJmYzk2ODBhYzVjN2IyYmE1ZGI5NGQxZGV8MTU5MDQwMTU4MzsgdXNlcm5hbWUtbG9jYWxob3N0LTg4ODg9XCIyfDE6MHwxMDoxNTkwNTc1MjIxfDIzOnVzZXJuYW1lLWxvY2FsaG9zdC04ODg4fDQ0Ok1EWTRaVEkwT0dWak1qSXhORGRsWWprME1tUTNPR0poTm1FMk5XTmtOR1E9fDY4MWNhZWExYWI0ZjNhMTBmMTYzNDI1MDMxOWY2Y2ZiOTMwMGRkYWU0MTcxYTcyNGE0MzI1NWNhYjhlZGE3OWFcIiIsICJYLUZvcndhcmRlZC1Db250ZXh0IjogIi9wcm94eS84Nzg3IiwgIlgtUHJveHljb250ZXh0cGF0aCI6ICIvcHJveHkvODc4NyJ9LCAiY29va2llcyI6IHsiX3hzcmYiOiAiMnw4YjkyMmUyY3xhMDc5NmRlYmZjOTY4MGFjNWM3YjJiYTVkYjk0ZDFkZXwxNTkwNDAxNTgzIiwgInVzZXJuYW1lLWxvY2FsaG9zdC04ODg4IjogIjJ8MTowfDEwOjE1OTA1NzUyMjF8MjM6dXNlcm5hbWUtbG9jYWxob3N0LTg4ODh8NDQ6TURZNFpUSTBPR1ZqTWpJeE5EZGxZamswTW1RM09HSmhObUUyTldOa05HUT18NjgxY2FlYTFhYjRmM2ExMGYxNjM0MjUwMzE5ZjZjZmI5MzAwZGRhZTQxNzFhNzI0YTQzMjU1Y2FiOGVkYTc5YSJ9fQ']
[I 14:30:05.139 LabApp] Trying to establish websocket connection to ws://localhost:8787/status/ws
[I 14:30:05.143 LabApp] Websocket connection established to ws://localhost:8787/status/ws
LabApp - ERROR - Uncaught exception
Traceback (most recent call last):
  File "/nobackup/earlacoa/miniconda3/envs/pangeo/lib/python3.6/site-packages/tornado/websocket.py", line 649, in _run_callback
    result = callback(*args, **kwargs)
  File "/nobackup/earlacoa/miniconda3/envs/pangeo/lib/python3.6/site-packages/tornado/websocket.py", line 1528, in on_message
    return self._on_message(message)
  File "/nobackup/earlacoa/miniconda3/envs/pangeo/lib/python3.6/site-packages/tornado/websocket.py", line 1534, in _on_message
    self._on_message_callback(message)
  File "/nobackup/earlacoa/miniconda3/envs/pangeo/lib/python3.6/site-packages/jupyter_server_proxy/handlers.py", line 296, in message_cb
    self.write_message(message, binary=isinstance(message, bytes))
  File "/nobackup/earlacoa/miniconda3/envs/pangeo/lib/python3.6/site-packages/tornado/websocket.py", line 339, in write_message
    raise WebSocketClosedError()
tornado.websocket.WebSocketClosedError
```
This websocket error doesn't occur every time, as sometimes the JupyterLab session on the command line dies with the following output, alongside the interactive session being killed:
```bash
[IPKernelApp] WARNING | Parent appears to have exited, shutting down.
```
I executed the same code in an IPython console and the same issue appears to happen, where the IPython console and the interactive session on the compute node both die when `.compute()` is executed. As shown below there were no outputted diagnostic from the IPython session, and the Dask workers have the same output as before:

```python
(pangeo) [earlacoa@d8s0b3.arc4 dask]$ ipython
Python 3.6.10 | packaged by conda-forge | (default, Apr 24 2020, 16:44:11) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.14.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: from dask_jobqueue import SGECluster                                                                                                                                                                                                                                                                                                                            

In [2]: from dask.distributed import Client                                                                                                                                                                                                                                                                                                                             

In [3]: import dask.array as da                                                                                                                                                                                                                                                                                                                                         

In [4]: cluster = SGECluster()                                                                                                                                                                                                                                                                                                                                          

In [5]: client = Client(cluster)                                                                                                                                                                                                                                                                                                                                        

In [6]: cluster.scale(jobs=50)                                                                                                                                                                                                                                                                                                                                          

In [7]: x = da.random.random((10000, 10000), chunks=(1000, 1000))                                                                                                                                                                                                                                                                                                       

In [8]: y = x + x.T                                                                                                                                                                                                                                                                                                                                                     

In [9]: y.compute()                                                                                                                                                                                                                                                                                                                                                     
(pangeo) [earlacoa@login1.arc4 scripts]$
```

**Comment by lesteve** [comment]
Does your SGE scheduler has something to say about the reason your job stopped (for example, maybe it used too much memory compared to the ressources you booked). On my SGE cluster I can do: `qacct -j <job_id>`

**Comment by lesteve** [comment]
An interesting thing that @lukeconibear said in the Dask-Jobqueue issue as well, is that it works fine for a small amount of worker (e.g. 5) but not with more workers (e.g. 50).

**Comment by lukeconibear** [comment]
Thank you for the very helpful suggestions.

> Does your SGE scheduler has something to say about the reason your job stopped (for example, maybe it used too much memory compared to the ressources you booked)? On my SGE cluster I can do: qacct -j <put_your_job_id_here> to get this kind of information.

Using `qacct -j <jobid>` I was able to see that the that interactive job's memory usage exceeded the memory allocated:

```bash
(pangeo) [earlacoa@login1.arc4 dask]$ qacct -j 529539
...
mem          1.061GBs
maxvmem      1.040GB
```
I assumed that as the only work being carried out was conducted on the Dask workers, that the resource usage of JupyterLab wouldn't be significant. This assumption held when there was only a small amount of work and number of workers (e.g. 5) or when the Dask dashboard was turned off. However, this assumption was incorrect with more workers or when using the Dask dashboard, and the insufficient memory resources caused the interactive job to be killed, which then caused the worker disconnections. 

When I request sufficient memory for the interactive job on the compute node, everything works.

I'm really grateful for your time and effort spent helping me here, and I apologise for any inconvenience and time wasting I've caused.

This issue can now be closed.
